### PR TITLE
Add SNES Macropad keyboard

### PR DESCRIPTION
**Pull request by JBarberU**
Adds keyboard SNES Macropad

## Description

Adds a new keyboard. It uses a shift register (the snes controller) for one "half" of the keyboard, and thus uses a custom_matrix. As best I could tell this was the best available approach, but would love input if there's a better way to combine a regular keyboard matrix alongside a shift register.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [x] Documentation (readme for the keyboard)

## Issues Fixed or Closed by This PR

None

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/snes_macropad/config.h
Comment (waffle87) [inline keyboards/snes_macropad/config.h:27]:
```suggestion
```

File: keyboards/snes_macropad/info.json
Comment (waffle87) [inline keyboards/snes_macropad/info.json:29]:
```suggestion
    },
    "rgblight": {
      "led_count": 12,
      "max_brightness": 80,
      "animations": {
        "alternating": true,
        "breathing": true,
        "christmas": true,
        "knight": true,
        "rainbow_mood": true,
        "rainbow_swirl": true,
        "rgb_test": true,
        "snake": true,
        "static_gradient": true,
        "twinkle": true
      }
    },
```

File: keyboards/snes_macropad/post_config.h
Comment (waffle87) [inline keyboards/snes_macropad/post_config.h:1]: This file should be removed.

File: keyboards/snes_macropad/readme.md
Comment (waffle87) [inline keyboards/snes_macropad/readme.md:3]: This image needs to be rehosted on Imgur.

File: keyboards/snes_macropad/rules.mk
Comment (waffle87) [inline keyboards/snes_macropad/rules.mk:4]:
```suggestion
```

File: keyboards/snes_macropad/rules.mk
Comment (waffle87) [inline keyboards/snes_macropad/rules.mk:7]: Please place this in an `halconf.h` file.
```suggestion
```

File: keyboards/snes_macropad/snes_macropad.c
Comment (waffle87) [inline keyboards/snes_macropad/snes_macropad.c:6]:
```suggestion
#include "quantum.h"
```

File: keyboards/snes_macropad/snes_macropad.c
Comment (waffle87) [inline keyboards/snes_macropad/snes_macropad.c:11]:
```suggestion
```

File: keyboards/snes_macropad/snes_macropad.c
Comment (waffle87) [inline keyboards/snes_macropad/snes_macropad.c:22]: This should be placed in a `snes_macropad.h` header file, or the functions in this file should be rearranged accordingly.
```suggestion
```

File: keyboards/snes_macropad/snes_macropad.c
Comment (waffle87) [inline keyboards/snes_macropad/snes_macropad.c:130]:
```suggestion
void keyboard_post_init_kb(void) {
    rgblight_enable_noeeprom();
    rgblight_sethsv_noeeprom(HSV_MAGENTA);
    rgblight_mode_noeeprom(RGBLIGHT_MODE_RAINBOW_SWIRL);
    keyboard_post_init_user();
}
```

File: keyboards/snes_macropad/snes_macropad.c
Comment (waffle87) [inline keyboards/snes_macropad/snes_macropad.c:23]:
```suggestion
bool oled_task_kb(void) {
    if (!oled_task_user()) {
        return false;
    }
```

File: keyboards/snes_macropad/info.json
Comment (waffle87) [inline keyboards/snes_macropad/info.json:13]:
```suggestion
        "nkro": true,
        "rgblight": true,
        "oled": true
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

**Comment by JBarberU** [comment]
Hey, thanks for the review!

I've pushed two commits addressing the issues, hopefully it looks better now.

File: keyboards/snes_macropad/config.h
Comment (waffle87) [inline keyboards/snes_macropad/config.h:10]: This is the default.
```suggestion
```

**Comment by waffle87** [comment]
It also compiles locally for me; I've a feeling this is an issue with the CI. Nothing you need to worry about.

**Review by waffle87 — APPROVED** [review]

**Comment by tzarc** [comment]
> Hmm, seems like that last commit failed to build properly. I tried a local build, which is working fine. Looking at logs it appears to be failing at creating the .uf2 files. Is there a way to trigger a CI rebuild, or might there be an actual problem that I need to dig into?

This one's on me; I'd moved our CI cluster to different hardware and the volume mounts were marked as `noexec`. We need to execute scripts to turn things into a `*.uf2`...

I've rectified and it's re-running now.

**Comment by JBarberU** [comment]
Hey, just want to make sure I'm reading things correctly that I'm still waiting for another review before this change can be merged?

**Comment by waffle87** [comment]
Correct, nothing you need to do right now.

File: keyboards/snes_macropad/info.json
Comment (drashna) [inline keyboards/snes_macropad/info.json:21]: just a heads up, you could specify the number of columns and rows instead of adding the `NO_PIN`s here.

File: keyboards/snes_macropad/keymaps/default/keymap.c
Comment (drashna) [inline keyboards/snes_macropad/keymaps/default/keymap.c:9]:
```suggestion
    L_RGB,
```

File: keyboards/snes_macropad/snes_macropad.c
Comment (drashna) [inline keyboards/snes_macropad/snes_macropad.c:91]: just a heads up, on develop (which is to be merged at the end of the month... `oled_render_dirty(true)` will handle this without the extra code here.

File: keyboards/snes_macropad/snes_macropad.c
Comment (drashna) [inline keyboards/snes_macropad/snes_macropad.c:50]: Since this is only 0-10 (less)... you could just call: 
```suggestion
    oled_write('0' + last_row, false);
```

That saves a bunch of code and calculations. And greatly simplifies the code

File: keyboards/snes_macropad/snes_macropad.c
Comment (drashna) [inline keyboards/snes_macropad/snes_macropad.c:77]:
```suggestion
    oled_advance_page(true);
```

File: keyboards/snes_macropad/snes_macropad.c
Comment (JBarberU) [reply keyboards/snes_macropad/snes_macropad.c]: Had to use oled_write_char instead of oled_write, but otherwise worked like a charm!

File: keyboards/snes_macropad/snes_macropad.c
Comment (JBarberU) [reply keyboards/snes_macropad/snes_macropad.c:90]: Added a todo with a commented out implementation. Wasn't sure if there was a better way?

File: keyboards/snes_macropad/snes_macropad.c
Comment (drashna) [inline keyboards/snes_macropad/snes_macropad.c:120]: Another thing to note here, is that `shutdown_user` is called by `QK_BOOT`.   Though, the caveats are that it currently has no way to differentiate between rebooting and jumping to bootloader, without adding manual tracking.  Also, you'd need to use `__attribute__((weak)) void shutdown_user(void)` since there is no kb version, currently.

**Review by drashna — APPROVED** [review]

File: keyboards/snes_macropad/snes_macropad.c
Comment (drashna) [reply keyboards/snes_macropad/snes_macropad.c:90]: Yeah, the newer function was added explicitly for this sort of use case in mind.  Which is why I mentioned it.

File: keyboards/snes_macropad/snes_macropad.c
Comment (JBarberU) [reply keyboards/snes_macropad/snes_macropad.c:120]: Sounds like the current implementation is simpler, or do you think I should change it? 

Either way I'm thinking it might make sense to mark it weak, to allow keymaps to override its behavior. But I'm currently putting that in the future improvements bucket.

Comment (drashna) [reply keyboards/snes_macropad/snes_macropad.c:120]: For now, this implementation is fine.  If you wanted to change things when rebooting, not just jumping to bootloader, then shutdown_user would be better.  But this works just fine.

**Pull request merged**
